### PR TITLE
feat(PPL Lint) Context-fed rules, hover card fixes, and a headless lint API for cross-repo CI

### PR DESCRIPTION
**Pull request by Hanyu-W**
### Description

**1. Three Context-Fed Rules**

- `enabled-false-object` (warning) — a reference into an object mapped `enabled: false`. Verified on 3.7: returns `null` with HTTP 200 and no error. Invisible without tooling, since `_field_caps` strips the attribute (which is also why `typeMap` can't detect it).
- `wildcard-source-zero-match` (info) — a `source=` wildcard matching no visible index.
-  `rex-scan-cost` (info) — Informs users that `rex`/`parse`/`grok` run their pattern over a source field for every input row, which can be slow on large indexes. Gives the user a suggestion to pre-filter their index to improve performance. 

**2. Hover card fixes**

- The Hover Card now has a 600 ms grace timer before hiding when moving the cursor off, making the hover card easier to interact with.
- One exported `ruleIdOf` replaces two private copies in the code-action and hover providers.
- Hover card descriptions have been simplified and made easier to understand for PPL beginners. The card now shows the rule's message, a `Fix` line, a quick-fix preview when one exists, and a `Learn more` link — the previous per-fact detail line (`hoverFacts`) has been removed, and with it the now-dead machinery that fed it (see review follow-up below).

** 3. Node-safe headless lint API** — extracts the bundle deserializer and the parse-and-lint core so the `opensearch-project/sql` CI can run OSD's production detectors against a candidate grammar bundle, without launching OSD, Monaco, or a browser. `runtime_lint.ts` now imports that shared core instead of keeping its own duplicate copies. Needs no new `@osd/monaco` export.

**4. Feature-usage telemetry** — wires PPL lint usage tracking through the Neo `core.telemetry` path (`core.telemetry.getPluginRecorder`), so we can answer "is anyone using PPL lint?". The recorder auto-stamps the `queryEnhancements` source and no-ops in public OSS (no `TelemetryProvider` is registered there).

`@osd/monaco` stays telemetry-API-agnostic: a new `lint/telemetry` module exposes `registerPPLLintTelemetry(sink)` (bridge-injection, mirroring `lint_bridge`) and `emitPPLLintTelemetry({ name, data })`. The engine emits four events, and the `query_enhancements` plugin forwards them to the recorder in `start()`. Events are deduped so they measure distinct interactions, not mouse travel (Monaco re-invokes `provideHover` per character and auto-fires `provideCodeActions` on every cursor move):

- `ppl_lint_diagnostic_shown` (once per rule per lint pass)
- `ppl_lint_hover_shown` (once per marker per pass)
- `ppl_lint_quickfix_offered` (once per marker per pass)
- `ppl_lint_quickfix_clicked` (via a command on the quick-fix action; Monaco applies the edit, then runs it)

### Issues Resolved

Supports https://github.com/opensearch-project/sql/issues/5405.

## Screenshot

**1. `enabled-false-object`** — a reference into an `enabled: false` object; the card explains the field is stored but not searchable.

enabled false object

**2. `wildcard-source-zero-match`** — a `source=` wildcard matching no visible index.

wildcard source zero match

**3. The `rex-scan-cost` rule** — flags the analyzed-text scan on the `raw_log` field.

rex scan cost

## Testing the changes

```
yarn test:jest packages/osd-monaco/src/ppl src/plugins/data/public/ppl_lint \
  src/plugins/data/public/antlr/opensearch_ppl src/plugins/data/public/ui/query_editor \
  src/plugins/explore/public/components/query_panel src/plugins/query_enhancements
```

2125 tests pass across 160 suites; `yarn typecheck` adds no new errors (verified with `tsc -b --force`, since the cached run reports success without compiling).

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuali925** [comment]
### Code review

Found 3 issues:

1. `enabled-false-object` never fires for nested disabled objects. The producer `collectDisabledObjectFields` emits full dotted names for nested `enabled:false` objects (its own test asserts `['outer.inner']`), but the detector only matches the first path segment, so a reference like `outer.inner.deep` computes `root = 'outer'`, `disabled.has('outer')` is false, and the rule silently never fires. Only top-level cases are tested. A prefix check (`path === d || path.startsWith(d + '.')`) would match the producer.

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/f56a45be5c8831bdd118814a6af90b1381b61c1c/packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts#L37-L41

2. Comments claim the hover card renders `hoverFacts`, but this PR removed that rendering. `renderHoverCard` now accepts only `severityLabel`, `message`, `docUrl`, `howToFix`, `fixText`, and `hover_provider.ts` no longer reads `hoverFacts` — yet both new rules carry comments asserting the card renders these facts, and in the wildcard rule this makes `nearbyCandidates()` ("did you mean…") dead computation.

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/f56a45be5c8831bdd118814a6af90b1381b61c1c/packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts#L49-L51

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/f56a45be5c8831bdd118814a6af90b1381b61c1c/packages/osd-monaco/src/ppl/lint/rules/wildcard_source_zero_match.ts#L94-L100

3. `enabled-false-object` is `enabled: true` with an empty `appliesTo`, so it fires on every engine/version. The rule's source says it was verified on OpenSearch 3.7 (Calcite), but the catalog entry has no `minVersion`/`engine` gate. Prior review on #12394 (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/12394#discussion_r3648098624) required this class of context-dependent mapping rule to be gated to the validated surface, adding `minVersion`/`engine: calcite` to `agg-on-text`, `type-mismatch-numeric`, and `flat-object-subfield`. This rule has the same shape.

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/f56a45be5c8831bdd118814a6af90b1381b61c1c/packages/osd-monaco/src/ppl/lint/rules_catalog.json#L195-L205

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by Hanyu-W** [comment]
Thanks @joshuali925 for the review. All three issues are addressed in dd56050.

  1. Nested disabled objects: Updated the detector to match full dotted paths using path === d || path.startsWith(d + '.'), consistent with the producer. Added nested-path coverage and a sibling-prefix negative case (log vs. logger.field).

  2. Dead hoverFacts: The card simplification had removed its only consumers, leaving every rule’s hoverFacts as dead writes. Removed the facility end to end, including the registry, language population, diagnostic attachment, candidate lookup, and all per-rule attachments. markerFixKey now imports from the canonical fix registry, and the explain-fix path remains unchanged. git grep hoverFacts packages/osd-monaco/src/ppl returns no matches.

  3. Ungated disabled-object rule: Added { minVersion: "3.7.0", engine: "calcite" }, matching the related rules from the #12394 review. 
   
 I also fixed a pre-existing test gap: three language_*.test.ts mocks lacked monaco.editor.registerCommand, which is required by the telemetry quick-fix command and caused those suites to fail independently of this change. Please re-review when you get the chance. Thanks!

File: packages/osd-monaco/src/ppl/lint/pattern_literal.ts [hunk 1]
```diff
@@ -47,3 +47,179 @@ export function findPatternLiteral(
...
+ * Steps (any may abort with `undefined`, always the safe default):
+ *   1. bail on a top-level alternation anywhere;
+ *   2. accumulate the leading literal run, stopping at the first metacharacter
+ *      (an anchor/metaclass/group/quantifier as the first char yields an empty
+ *      run and thus no token);
+ *   3. split the run into analyzer-like tokens and keep only a clean ASCII token
+ *      that is not pure-numeric, not a stopword, and at least MIN_TOKEN_LENGTH;
+ *      return the longest survivor.
+ */
+export function leadingLiteralToken(decoded: string): string | undefined {
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/pattern_literal.ts]: `leadingLiteralToken`, `decodePatternLiteral`, `hasTopLevelAlternation`, `STOPWORDS` and `MIN_TOKEN_LENGTH` are exported and covered by the new 130-line `__tests__/pattern_literal.test.ts`, but nothing calls them. `rex_scan_cost.ts` imports only from `rule_index`, `pipeline_shape` and `range_utils`; I grepped the rest of `packages/` and `src/` and found no other reference.

The PR description explains why — the prefilter-token suggestion moved out of the message and onto the static catalog `howToFix` line. So this is the machinery that fed the removed feature, which is the same category of thing item 2 of your review follow-up deletes for `hoverFacts`.

Could this come out too, or does `rex-scan-cost` still intend to name a candidate token? ~185 lines of production code plus its tests is a lot to carry for a future caller.

File: src/plugins/data/public/ppl_lint/lint_context_builder.ts [hunk 2]
```diff
@@ -132,7 +136,11 @@ export function buildPPLLintContext(
     isCalcite,
     fields: cacheMatchesDataset ? lintFields.fields : undefined,
     typeMap: cacheMatchesDataset ? lintFields.typeMap : undefined,
+    disabledObjectFields: cacheMatchesDataset ? lintFields.disabledObjectFields : undefined,
     selectedSourcePattern: cacheMatchesDataset ? lintFields.selectedSourcePattern : undefined,
+    // Cluster-wide rather than dataset-scoped, so it is not gated on the
+    // dataset-identity check the field metadata uses.
+    visibleIndices: lintFields.visibleIndices,
```
Comment (RyanL1997) [inline src/plugins/data/public/ppl_lint/lint_context_builder.ts]: `fields`, `typeMap` and `disabledObjectFields` are all gated on `cacheMatchesDataset`, but `visibleIndices` is passed through ungated. I follow the reasoning in the comment — it's cluster-wide, not dataset-scoped — but the cache object is only replaced wholesale once the async fetch resolves. So during the window right after a dataset switch, `lintFields.visibleIndices` still holds the *previous* dataset's list.

Within one cluster that's harmless. Across an MDS data-source switch it's a different cluster's index names, so `wildcard-source-zero-match` can fire against the wrong cluster and tell the user their pattern matches nothing when it matches fine.

Gating on `dataSourceId` equality (rather than full dataset identity) would keep the cluster-wide semantics while closing that window.

File: packages/osd-monaco/src/ppl/language.ts [hunk 3]
```diff
@@ -341,6 +345,27 @@ const processLintHighlighting = (model: monaco.editor.IModel): void => {
         generation,
         lintResult.diagnostics
       );
+
+      // A new marker set is a fresh opportunity: re-arm the per-pass hover /
+      // quick-fix-offered dedup so interactions with this pass's markers count
+      // again (they were deduped within the previous pass).
+      resetPPLLintTelemetryDedup(model);
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/language.ts]: `resetPPLLintTelemetryDedup(model)` runs on every new marker set, and a marker set is produced per debounced keystroke — so `diagnostic_shown` is deduped *within* a pass but re-arms on every pass.

That means a user typing a 60-character query with one `division-by-zero` warning emits dozens of `ppl_lint_diagnostic_shown` events. The metric then measures "how much do people type near a lint finding," not "is anyone using PPL lint" — which is the question the PR sets out to answer.

Is that the intended denominator? If the goal is distinct findings surfaced, the dedup probably wants to survive across passes for an unchanged marker (keyed on the marker set's content rather than reset on every apply). Also worth a sanity check on the volume this pushes at the recorder.

File: packages/osd-monaco/src/ppl/lint/catalog.ts [hunk 4]
```diff
@@ -40,6 +40,7 @@ export function validateCatalogEntry(value: unknown): CatalogEntry | null {
     typeof candidate.severity !== 'string' ||
     !VALID_SEVERITIES.has(candidate.severity) ||
     typeof candidate.message !== 'string' ||
+    typeof candidate.howToFix !== 'string' ||
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/catalog.ts]: This makes `howToFix` mandatory — an entry missing it is rejected outright and the rule silently vanishes rather than rendering without a Fix line.

That's fine for the in-repo `rules_catalog.json`, which this PR updates in lockstep. Can you confirm nothing else feeds entries through `validateCatalogEntry`? I'm specifically thinking about catalogs supplied by the runtime grammar bundle, where an older producer wouldn't know about the field and would lose every rule at once. If that path exists, defaulting to `''` and skipping the Fix line would fail a lot softer.

File: packages/osd-monaco/src/ppl/lint/hover/hover_card.ts [hunk 5]
```diff
@@ -55,115 +58,45 @@ function code(text: string): string {
...
-
-  if (content) {
+  // Every known rule gives the user a concrete next action, whether or not an
+  // automatic edit can be offered safely.
+  if (howToFix) {
     lines.push('');
-    lines.push(
-      `**Why ${severityLabel.toLowerCase()}** — ${FAILURE_CLASS_EXPLAINER[content.failureClass]}`
-    );
+    lines.push(`**Fix** — ${howToFix}`);
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/hover/hover_card.ts:82]: `message` is run through `escapeInline` two lines up, but `howToFix` is interpolated raw. I assume that's deliberate — the catalog text uses backticks for inline code and would render as literal backticks if escaped — but the asymmetry deserves a comment next to the careful `escapeInline` note above, so nobody "fixes" it later.

Separate, more substantive: this reads `howToFix` from `getCatalogEntryById(ruleId)` (the static bundled catalog), while `marker.message` reflects the config after `mergeConfig` applied `bundleOverrides`. So a bundle that overrides `howToFix` gets the overridden text ignored and the built-in shown instead — a silent divergence between the two halves of the same card. Worth either plumbing the merged config through or documenting that `howToFix` is deliberately not overridable.

File: packages/osd-monaco/src/ppl/lint/rules/wildcard_source_zero_match.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import { Diagnostic } from '../diagnostic';
+import { Detector } from '../types';
+import { classifyTopLevelSource } from '../top_level_source';
+
+/**
+ * Convert a PPL wildcard pattern (`*` matches any run of characters) into a
+ * RegExp anchored to the full string.
+ */
+function wildcardToRegExp(pattern: string): RegExp {
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/rules/wildcard_source_zero_match.ts:14]: This handles `*` only. A few shapes OpenSearch resolves that would fall through to a false "matches no known index":

- comma-separated multi-target: `source=logs-*,metrics-*`
- exclusions: `source=logs-*,-logs-2023*`
- date-math: `source=<logs-{now/d}>`

It's `info` severity so the blast radius is small, but a linter that cries wolf on a valid query loses trust quickly, and these are the exact patterns a user with a wildcard source is likely to reach for. Bailing out (no diagnostic) when the pattern contains `,`, a leading `-`, or `<`/`>` would be a cheap way to stay conservative.

Also worth confirming: does `resolve_index/*` include hidden and system indices? If not, a legitimate `source=.something*` would false-fire.

File: packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  const diagnostics: Diagnostic[] = [];
+
+  for (const node of collectDottedPathNodes(tree, ruleNameToIndex)) {
+    const path = node.getText();
+    // A disabled object may itself be dotted (e.g. `outer.inner`), which is the
+    // form the producer (`collectDisabledObjectFields`) emits. Match the full
+    // path or any subfield beneath it; the `d + '.'` boundary keeps `log` from
+    // matching an unrelated `logger.field`.
+    const disabledRoot = [...disabled].find((d) => path === d || path.startsWith(d + '.'));
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts]: `[...disabled]` allocates a fresh array from the Set on every iteration of the loop over dotted-path nodes. Hoisting it above the `for` would make this O(n·m) without the per-node allocation — cheap to fix and this runs on every lint pass.

File: src/plugins/data/public/ppl_lint/visible_indices.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      { query }
+    );
+    if (!response) {
+      return [];
+    }
+    const names: string[] = [];
+    response.indices?.forEach((i) => names.push(i.name));
+    response.aliases?.forEach((a) => names.push(a.name));
+    response.data_streams?.forEach((d) => names.push(d.name));
+    if (names.length > MAX_VISIBLE_INDICES) {
```
Comment (RyanL1997) [inline src/plugins/data/public/ppl_lint/visible_indices.ts]: Worth being precise in the doc comment: this cap bounds what's handed to the lint engine, but it's applied *after* the full `resolve_index/*` payload has been downloaded and parsed. On a cluster with tens of thousands of indices the network and server cost is paid in full and then discarded.

It's per dataset change rather than per keystroke, so it's not alarming — but the comment currently reads as though the cap avoids the cost, and someone tuning this later would be misled. A `?expand_wildcards` / per-pattern narrowing query is the real fix if it ever becomes a problem.

File: packages/osd-monaco/src/ppl/lint/rules_catalog.json [hunk 9]
```diff
@@ -168,12 +182,48 @@
...
+    "enabled": true,
+    "severity": "info",
+    "message": "Wildcard source pattern matches no known index.",
+    "howToFix": "Correct the `source=` pattern or choose an existing index.",
+    "docUrl": "https://docs.opensearch.org/latest/sql-and-ppl/ppl/commands/search/",
+    "needsContext": true,
+    "appliesTo": {}
+  },
+  {
+    "id": "rex-scan-cost",
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/rules_catalog.json:218]: The catalog `message` here hardcodes "rex", but `rexScanCostDetector` overrides it per-command via `scanCostMessage(keyword, field)`, so a `parse` or `grok` finding gets the right keyword at runtime and this string is never shown.

Harmless today, but it's a latent mismatch for anything that reads catalog text directly (docs generation, i18n extraction, the doc-links snapshot). Consider making it keyword-neutral — e.g. "Pattern extraction runs the pattern against every input row from the text field, even when it finds no match."

File: packages/osd-monaco/src/ppl/lint/hover/hover_card.ts [hunk 10]
```diff
@@ -55,115 +58,45 @@ function code(text: string): string {
...
+ */
 export function renderHoverCard(input: HoverCardInput): string {
-  const { ruleId, severityLabel, message, docUrl, content, facts } = input;
+  const { severityLabel, message, docUrl, howToFix, fixText } = input;
   const lines: string[] = [];
 
-  lines.push(`${SEVERITY_GLYPH[severityLabel]} **${escapeInline(ruleId)}** · ${severityLabel}`);
+  // The rule id remains on the marker for lookup and support diagnostics, but it
+  // is implementation detail rather than the card's headline.
+  lines.push(`${SEVERITY_GLYPH[severityLabel]} **${severityLabel}**`);
```
Comment (RyanL1997) [inline packages/osd-monaco/src/ppl/lint/hover/hover_card.ts]: The card now leads with just the severity glyph and label; the rule id is gone.

The id is how a user turns off a rule they find noisy (Advanced Settings keys off it) and how support correlates a screenshot to a rule. Removing it from the card removes the only place it surfaces in the UI, so a user who wants to silence `rex-scan-cost` has no way to learn what it's called.

I agree it shouldn't be the headline — but could it survive somewhere subtler? Appended to the severity line in a dimmer form, or folded into the `Learn more` row, would keep it discoverable without competing with the message.

File: src/plugins/data/public/ppl_lint/visible_indices.ts (hunk 8 above)
Comment (RyanL1997) [reply src/plugins/data/public/ppl_lint/visible_indices.ts]: I suggest we do a test with a real-world like data load

File: packages/osd-monaco/src/ppl/lint/rules_catalog.json (hunk 9 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/rules_catalog.json:218]: Agreed. The detector already emits command-specific messages for rex, parse, and grok, but the catalog fallback is misleading. I will make the catalog message command-neutral and add coverage for that copy while retaining the existing per-command diagnostic assertions.

File: packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts (hunk 7 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/rules/enabled_false_object.ts]: Agreed. I will replace the per-node spread with direct Set iteration that short-circuits on the first matching root. I will also add an overlapping-root test to preserve the one-diagnostic-per-reference behavior.

File: packages/osd-monaco/src/ppl/lint/catalog.ts (hunk 4 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/catalog.ts]: Confirmed: runtime grammar bundles do not carry lint catalog entries. Both browser runtime lint and the headless API use OSD's bundled rules_catalog.json, and Advanced Settings supplies only per-rule overrides. An older grammar producer therefore cannot omit howToFix or make rules disappear. I will keep the bundled-catalog field required, add a comment that records this ownership boundary, and retain test coverage that every shipping entry has non-empty fix guidance.

File: packages/osd-monaco/src/ppl/lint/pattern_literal.ts (hunk 1 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/pattern_literal.ts]: Good catch. Removed the unused decodePatternLiteral, leadingLiteralToken, hasTopLevelAlternation, stopword, and token-selection machinery, along with the test file that only covered that dead path. findPatternLiteral remains because invalid-capture-group-name still uses it, and that rule's tests continue to cover the retained parse-tree lookup.

File: src/plugins/data/public/ppl_lint/lint_context_builder.ts (hunk 2 above)
Comment (Hanyu-W) [reply src/plugins/data/public/ppl_lint/lint_context_builder.ts]: Good catch. visibleIndices is now gated by cached versus active dataSourceId, so names from the previous MDS cluster are suppressed while the new request is pending. The cache remains reusable across datasets on the same cluster because the inventory is cluster-wide. I added coverage for cross-MDS suppression, same-cluster reuse, and the local cluster case where both IDs are undefined.

File: packages/osd-monaco/src/ppl/language.ts (hunk 3 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/language.ts]: Agreed, per-pass emission was the wrong denominator. Telemetry now reconciles model-scoped active state across accepted lint passes. diagnostic_shown fires when a rule transitions from absent to present; hover and quick-fix-offered remain deduped for each exact active marker fingerprint. A finding is re-armed only after it disappears from an accepted state and later returns, while quickfix_clicked still records every invocation.

The state is bounded to active findings in a WeakMap and is explicitly cleared when lint is disabled, the model leaves PPL, the model is disposed, or language support is torn down. A regression test runs 60 accepted passes with the same diagnostic and verifies that only one exposure event is emitted; removal followed by return starts a second episode.

File: packages/osd-monaco/src/ppl/lint/hover/hover_card.ts (hunk 5 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/hover/hover_card.ts:82]: Agreed on both ambiguities. howToFix is repository-authored Markdown, so its inline-code backticks are intentionally preserved; the completed hover block still uses isTrusted: false. I added a comment at the interpolation point to make that trust boundary explicit.

I also traced and documented the override path. Runtime grammar bundles do not supply presentation metadata, and the Advanced Settings adapter intentionally emits only enabled and severity. The broader BundleRuleOverrides type remains available to engine and headless callers, but Monaco hover guidance deliberately comes from the bundled catalog rather than execution-time overrides. Tests verify that settings-provided presentation fields are ignored and that catalog inline-code Markdown still renders

File: src/plugins/data/public/ppl_lint/visible_indices.ts (hunk 8 above)
Comment (Hanyu-W) [reply src/plugins/data/public/ppl_lint/visible_indices.ts]: Did a real-world like data load, scaling _resolve/index/* from 5 → 20,000 names: 505 names = 24 KiB / 1.3 ms; 5,000 = 412 KiB / 7.6 ms; 20,000 = 1.7 MB / 33 ms + 15 ms parse. So yes, at 20,000 names ~48 ms and ~1 MB get downloaded and discarded.

However, fetchVisibleIndices runs from a useEffect keyed on dataset/dataSource — once per dataset switch, not per keystroke — already inside the existing Promise.all, with the result cached in lintFieldsRef. Avoiding the waste needs an extra probe request that every user pays on every switch, to save ~48 ms for people on >5,000-index clusters.

File: packages/osd-monaco/src/ppl/lint/hover/hover_card.ts (hunk 10 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/hover/hover_card.ts]: Agreed, removing it from the severity Line, as it already is shown in the message

File: packages/osd-monaco/src/ppl/lint/rules/wildcard_source_zero_match.ts (hunk 6 above)
Comment (Hanyu-W) [reply packages/osd-monaco/src/ppl/lint/rules/wildcard_source_zero_match.ts:14]: Confirmed. The detector now diagnoses only simple local wildcard patterns and fails open for comma-separated targets, exclusions, date math, cross-cluster sources, unsupported punctuation, and dot-prefixed targets.

Bare /_resolve/index/* defaults to expand_wildcards=open, so hidden indices can be omitted. The inventory request now sends expand_wildcards=all; however, that does not bypass system-index permissions, and the legacy compatibility transport may ignore it. Dot-prefixed patterns therefore still fail open to avoid false warnings.

  I added detector, runtime grammar, request-forwarding, and route coverage for these cases.

**Review by mengweieric — APPROVED** [review]

**Review by joshuali925 — APPROVED** [review]
